libraries/networking/src/udt/CongestionControl.cpp

```diff
@@ -0,0 +1,205 @@
+    if (_loss) {
+        _loss = false;
+        return;
+    }
+    
+    double capacitySpeedDelta = (_bandwidth - USECS_PER_SECOND / _packetSendPeriod);
+    
+    // UDT uses what they call DAIMD - additive increase multiplicative decrease with decreasing increases
+    // This factor is a protocol parameter that is part of the DAIMD algorithim
+    static const int AIMD_DECREASING_INCREASE_FACTOR = 9;
+    
+    if ((_packetSendPeriod > _lastDecreasePeriod) && ((_bandwidth / AIMD_DECREASING_INCREASE_FACTOR) < capacitySpeedDelta)) {
+        capacitySpeedDelta = _bandwidth / AIMD_DECREASING_INCREASE_FACTOR;
+    }
+    
+    if (capacitySpeedDelta <= 0) {
+        increase = minimumIncrease;
+    } else {
+        // use UDTs DAIMD algorithm to figure out what the send period increase factor should be
+        
+        // inc = max(10 ^ ceil(log10(B * MSS * 8 ) * Beta / MSS, minimumIncrease)
+        // B = estimated link capacity
+        // Beta = 1.5 * 10^(-6)
+        
+        static const double BETA = 0.0000015;
+        static const double BITS_PER_BYTE = 8.0;
+        
+        increase = pow(10.0, ceil(log10(capacitySpeedDelta * _mss * BITS_PER_BYTE))) * BETA / _mss;
+        
+        if (increase < minimumIncrease) {
+            increase = minimumIncrease;
+        }
+    }
+    
+    setPacketSendPeriod((_packetSendPeriod * synInterval()) / (_packetSendPeriod * increase + synInterval()));
+}
+
+void DefaultCC::onLoss(SequenceNumber rangeStart, SequenceNumber rangeEnd) {
+    // stop the slow start if we haven't yet
+    if (_slowStart) {
+        stopSlowStart();
+        
+        // if the change to send rate was driven by a known receive rate, then we don't continue with the decrease
+        if (_receiveRate > 0) {
+            return;
+        }
+    }
+    
+    _loss = true;
+    
+    static const double INTER_PACKET_ARRIVAL_INCREASE = 1.125;
+    static const int MAX_DECREASES_PER_CONGESTION_EPOCH = 5;
+    
+    // check if this NAK starts a new congestion period - this will be the case if the
+    // NAK received occured for a packet sent after the last decrease
+    if (rangeStart > _lastDecreaseMaxSeq) {
+        _lastDecreasePeriod = _packetSendPeriod;
+        
+        _packetSendPeriod = ceil(_packetSendPeriod * INTER_PACKET_ARRIVAL_INCREASE);
+        
+        // use EWMA to update the average number of NAKs per congestion
+        static const double NAK_EWMA_ALPHA = 0.125;
+        _avgNAKNum = (int)ceil(_avgNAKNum * (1 - NAK_EWMA_ALPHA) + _nakCount * NAK_EWMA_ALPHA);
+        
+        // update the count of NAKs and count of decreases in this interval
+        _nakCount = 1;
+        _decreaseCount = 1;
+        
+        _lastDecreaseMaxSeq = _sendCurrSeqNum;
+        
+        // avoid synchronous rate decrease across connections using randomization
+        srand((unsigned) _lastDecreaseMaxSeq);
+        _randomDecreaseThreshold = (int) ceil(_avgNAKNum * (double(rand()) / RAND_MAX));
```

Comment:
I think you're half joking but honestly I'd like to fix this to use the fast c++11 pseudo-random